File: cpp/tests/test_top_addressing.cpp
// 动态 name server（DefaultTopAddressing）单测 —— 本地 mock HTTP server，不依赖外网。
//
// Java 语义锚点与 Python 侧 test_top_addressing.py 同源：
//   * WS 地址 / unitName / para 的 URL 拼装规则；
//   * clearNewLine（trim 后截断到第一个 \r 或 \n）；
//   * 200 → 地址串；非 200 / 连接失败 → 空串；
//   * fetchAndApply：地址**变化才应用**；
//   * MQClientInstance：配了静态地址不 fetch；空地址 fetch 一次并应用；取不到报错；
//     只有走动态取址时才起 10s/2min 的刷新线程（Java scheduleAtFixedRate）；
//   * configureFromEnv(ROCKETMQ_NAMESRV_DOMAIN)：实例构造时自己装配 wsAddr + unitName。
#include <atomic>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#if defined(_WIN32)
#include <winsock2.h>
#include <ws2tcpip.h>
#else
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>
#endif

#include "rocketmq/client/mq_client.h"
#include "rocketmq/common/net_compat.h"
#include "rocketmq/common/top_addressing.h"

using namespace rocketmq;

#if defined(_WIN32)
static const int kShutdownHow = SD_BOTH;
#else
static const int kShutdownHow = SHUT_RDWR;
#endif

namespace {

int fails = 0;
int checks = 0;

void expect(bool ok, const std::string& name, const std::string& detail = "") {
    ++checks;
    if (!ok) {
        ++fails;
        std::printf("FAIL %s %s\n", name.c_str(), detail.c_str());
    }
}

void expectEq(const std::string& actual, const std::string& expected, const std::string& name) {
    ++checks;
    if (actual != expected) {
        ++fails;
        std::printf("FAIL %s (actual=[%s] expected=[%s])\n", name.c_str(), actual.c_str(),
                    expected.c_str());
    }
}

// 临时改环境变量、析构时还原（值设成空串即等价于"未设置"：实现按 env[0]=='\0' 判断）。
// ::setenv 是 POSIX 专有，MSVC 只有 _putenv_s。
class ScopedEnv {
 public:
    ScopedEnv(const char* name, const char* value) : name_(name) {
        const char* old = std::getenv(name);
        if (old != nullptr) oldValue_ = old;
        apply(name, value);
    }
    ~ScopedEnv() { apply(name_, hadOld() ? oldValue_.c_str() : ""); }

    ScopedEnv(const ScopedEnv&) = delete;
    ScopedEnv& operator=(const ScopedEnv&) = delete;

    bool hadOld() const { return !oldValue_.empty(); }

 private:
    void apply(const char* name, const char* value) {
#if defined(_WIN32)
        _putenv_s(name, value);
#else
        ::setenv(name, value, 1);
#endif
    }

    const char* name_;
    std::string oldValue_;
};

// ------------------------------------------------------------------ mock server

class MockAddrServer {
 public:
    MockAddrServer() {
        netcompat::ensureInitialized();
        listen_ = ::socket(AF_INET, SOCK_STREAM, 0);
        int one = 1;
        ::setsockopt(listen_, SOL_SOCKET, SO_REUSEADDR,
                     reinterpret_cast<const char*>(&one), sizeof(one));
        sockaddr_in addr{};
        addr.sin_family = AF_INET;
        addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
        addr.sin_port = 0;    // 系统分配
        ::bind(listen_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr));
        netcompat::socklen_type len = sizeof(addr);
        ::getsockname(listen_, reinterpret_cast<sockaddr*>(&addr), &len);
        port_ = ntohs(addr.sin_port);
        ::listen(listen_, 8);
        thread_ = std::thread([this]() { loop(); });
    }

    ~MockAddrServer() {
        running_ = false;
        ::shutdown(listen_, kShutdownHow);
        netcompat::closeSocket(listen_);
        thread_.join();
    }

    int port() const { return port_; }

    void setResponse(int status, const std::string& body) {
        std::lock_guard<std::mutex> lk(m_);
        status_ = status;
        body_ = body;
    }

    int hitCount() {
        std::lock_guard<std::mutex> lk(m_);
        return hits_;
    }

 private:
    void loop() {
        running_ = true;
        while (running_) {
            sockaddr_in peer{};
            netcompat::socklen_type len = sizeof(peer);
            netcompat::socket_t conn = ::accept(listen_, reinterpret_cast<sockaddr*>(&peer), &len);
            if (!netcompat::isValid(conn)) break;
            handle(conn);
            netcompat::closeSocket(conn);
        }
    }

    void handle(netcompat::socket_t conn) {
        char buf[4096] = {0};
        if (::recv(conn, buf, sizeof(buf) - 1, 0) <= 0) return;
        std::lock_guard<std::mutex> lk(m_);
        ++hits_;
        const std::string head = "HTTP/1.0 " + std::to_string(status_)
            + (status_ == 200 ? " OK" : " ERR")
            + "\r\nContent-Length: " + std::to_string(body_.size())
            + "\r\nConnection: close\r\n\r\n";
        const std::string resp = head + body_;
        ::send(conn, resp.data(), static_cast<int>(resp.size()), netcompat::sendFlags());
    }

    netcompat::socket_t listen_;
    int port_ = 0;
    std::thread thread_;
    std::atomic<bool> running_{false};
    std::mutex m_;
    int status_ = 200;
    std::string body_ = "127.0.0.1:9876";
    int hits_ = 0;
};

std::string topAddrDomain(const MockAddrServer& s) {
    return "127.0.0.1:" + std::to_string(s.port());
}

// C++ 构造函数第一个参数是 **wsAddr**（与 Python 的 domain= 关键字不同），
// 这里统一用 getWsAddr 生成完整 URL。
std::string topAddrWs(const MockAddrServer& s) {
    return DefaultTopAddressing::getWsAddr(topAddrDomain(s));
}

// ------------------------------------------------------------------ 用例

void testGetWsAddr() {
    expectEq(DefaultTopAddressing::getWsAddr("jmenv.tbsite.net"),
             "http://jmenv.tbsite.net:8080/rocketmq/nsaddr", "default port appended");
    expectEq(DefaultTopAddressing::getWsAddr("host:12345"),
             "http://host:12345/rocketmq/nsaddr", "domain with port skips default");
    expectEq(DefaultTopAddressing::getWsAddr("h", "mygrp"),
             "http://h:8080/rocketmq/mygrp", "custom subgroup");
}

void testBuildUrl() {
    DefaultTopAddressing plain("http://h:8080/rocketmq/nsaddr");
    expectEq(plain.buildUrl(), "http://h:8080/rocketmq/nsaddr", "plain url");

    DefaultTopAddressing withUnit("http://h:8080/rocketmq/nsaddr", "unitA");
    expectEq(withUnit.buildUrl(), "http://h:8080/rocketmq/nsaddr-unitA?nofix=1", "unit name");

    std::map<std::string, std::string> para;
    para["k"] = "v";
    DefaultTopAddressing unitPara("http://h:8080/rocketmq/nsaddr", "u", para);
    expectEq(unitPara.buildUrl(), "http://h:8080/rocketmq/nsaddr-u?nofix=1&k=v",
             "unit name with para");

    DefaultTopAddressing blankUnit("http://h:8080/rocketmq/nsaddr", "   ");
    expectEq(blankUnit.buildUrl(), "http://h:8080/rocketmq/nsaddr", "blank unit ignored");
}

void testClearNewLine() {
    expectEq(clearNewLine("  1.2.3.4:9876\r\nrest"), "1.2.3.4:9876", "cut at cr");
    expectEq(clearNewLine("a:9876\nb:9877"), "a:9876", "cut at lf");
    expectEq(clearNewLine("  a:9876  "), "a:9876", "trim only");
    expectEq(clearNewLine("\r\n"), "", "empty");
}

void testFetchBehavior(MockAddrServer& s) {
    s.setResponse(200, "10.0.0.1:9876;10.0.0.2:9876\nextra");
    DefaultTopAddressing ta(topAddrWs(s));
    expectEq(ta.fetchNsAddr(false), "10.0.0.1:9876;10.0.0.2:9876", "200 returns cleared body");

    s.setResponse(500, "ignored");
    expectEq(ta.fetchNsAddr(false), "", "non-200 returns empty");

    DefaultTopAddressing dead("http://127.0.0.1:1/rocketmq/nsaddr", std::string(),
                              std::map<std::string, std::string>(), /*timeoutMillis=*/300);
    expectEq(dead.fetchNsAddr(false), "", "connection error returns empty");
}

void testFetchAndApply(MockAddrServer& s) {
    s.setResponse(200, "127.0.0.1:9876");
    DefaultTopAddressing ta(topAddrWs(s));
    expectEq(ta.fetchAndApply(), "127.0.0.1:9876", "first change applied");
    expectEq(ta.fetchAndApply(), "", "same addr not applied again");
    s.setResponse(200, "10.0.0.9:9876");
    expectEq(ta.fetchAndApply(), "10.0.0.9:9876", "changed addr applied");
}

void testMqClientIntegration(MockAddrServer& s) {
    // 配了静态地址 → 不该问地址服务器（hits 用差值比较，前面用例已经打过它）
    const int baseline = s.hitCount();
    {
        MQClientInstance mqc("c@cppdyn", {"127.0.0.1:9876"});
        mqc.topAddressing().setWsAddr(topAddrWs(s));
        mqc.start();
        mqc.shutdown();
        expectEq(s.hitCount() == baseline ? "0" : "n", "0", "static addrs never fetch");
    }
    // 空地址 → start 时 fetch 一次并应用
    s.setResponse(200, "127.0.0.1:9876");
    {
        MQClientInstance mqc("c@cppdyn2", {});
        mqc.topAddressing().setWsAddr(topAddrWs(s));
        mqc.start();
        const std::vector<std::string> addrs = mqc.nameServerAddrs();
        expect(addrs.size() == 1 && addrs[0] == "127.0.0.1:9876",
               "empty addrs fetched at start", "size=" + std::to_string(addrs.size()));
        expectEq(s.hitCount() > 0 ? "y" : "n", "y", "address server hit");
        mqc.shutdown();
    }
    // 地址服务器取不到 → start 报错
    s.setResponse(500, "");
    {
        MQClientInstance mqc("c@cppdyn3", {});
        mqc.topAddressing().setWsAddr(topAddrWs(s));
        bool threw = false;
        try {
            mqc.start();
        } catch (const std::exception&) {
            threw = true;
        }
        expect(threw, "start fails when address server returns none");
    }
}

void testDefaultConstructedIsDisabled() {
    DefaultTopAddressing ta;
    expect(ta.wsAddr().empty(), "default ctor has no ws addr");
    expectEq(ta.fetchNsAddr(false), "", "disabled fetch returns empty");
}

// configureFromEnv：对应 Java MQClientAPIImpl 构造里的 `new DefaultTopAddressing(unitName)`
// —— 那个构造函数自己读域名（Java 是系统属性，本端口统一用环境变量
// ROCKETMQ_NAMESRV_DOMAIN，与 python/rust/dotnet 三端口一致）。
// 这条链路此前只能靠调用方手工 setWsAddr 才通，实例自己装配是缺的。
void testConfigureFromEnv(MockAddrServer& s) {
    const std::string key = "ROCKETMQ_NAMESRV_DOMAIN";
    // 1) 未配置 = 完全 no-op（不能让没设 env 的用户白等 3s 域名解析）
    {
        ScopedEnv off(key.c_str(), "");
        expect(!DefaultTopAddressing::isConfigured(), "env 未设置时 isConfigured=false");
        DefaultTopAddressing ta;
        ta.configureFromEnv("unitA");
        expect(ta.wsAddr().empty(), "env 未设置 → 不改 wsAddr");
        expect(ta.unitName().empty(), "env 未设置 → 不记 unitName");
        expectEq(ta.fetchNsAddr(false), "", "env 未设置时 fetch 仍是 no-op");
    }
    // 2) 配置了 domain：wsAddr 按 MixAll.getWSAddr 拼，unitName 进 URL
    s.setResponse(200, "127.0.0.1:9876");
    {
        ScopedEnv on(key.c_str(), topAddrDomain(s).c_str());
        expect(DefaultTopAddressing::isConfigured(), "env 设置后 isConfigured=true");
        DefaultTopAddressing ta;
        ta.configureFromEnv("unitA");
        expectEq(ta.wsAddr(), topAddrWs(s), "configureFromEnv 拼出 wsAddr");
        expectEq(ta.unitName(), "unitA", "unitName 落到地址服务器配置");
        // domain 自带端口 ⇒ 不追加 :8080；unitName 段是 "-<unit>?nofix=1"
        expectEq(ta.buildUrl(), topAddrWs(s) + "-unitA?nofix=1", "env + unitName 的 URL");
        expectEq(ta.fetchNsAddr(false), "127.0.0.1:9876", "env 驱动的 fetch 取到地址");
    }
    // 3) MQClientInstance 构造时自己装配：facade 不用手工 setWsAddr
    {
        ScopedEnv on(key.c_str(), topAddrDomain(s).c_str());
        MQClientInstance mqc("c@cppenv", {}, 3000, 15000, false, "unitZ");
        expectEq(mqc.topAddressing().wsAddr(), topAddrWs(s), "实例从 env 装好 wsAddr");
        expectEq(mqc.topAddressing().unitName(), "unitZ", "实例把 unitName 传给取址配置");
        mqc.start();
        const std::vector<std::string> addrs = mqc.nameServerAddrs();
        expect(addrs.size() == 1 && addrs[0] == "127.0.0.1:9876",
               "env 驱动的实例 start 时取到地址", "addrs=" + std::to_string(addrs.size()));
        mqc.shutdown();
    }
    // 4) 没有 env 时实例保持关闭动态取址，也不碰地址服务器
    {
        ScopedEnv off(key.c_str(), "");
        MQClientInstance mqc("c@cppenv2", {"127.0.0.1:9876"});
        expect(mqc.topAddressing().wsAddr().empty(), "无 env 时实例 wsAddr 为空");
        const int before = s.hitCount();
        mqc.start();
        mqc.shutdown();
        expectEq(std::to_string(s.hitCount() - before), "0", "无 env 时不请求地址服务器");
    }
}

}  // namespace

int main() {
    testGetWsAddr();
    testBuildUrl();
    testClearNewLine();
    {
        MockAddrServer s;
        testFetchBehavior(s);
        testFetchAndApply(s);
        testMqClientIntegration(s);
    }
    testDefaultConstructedIsDisabled();
    {
        MockAddrServer s;
        testConfigureFromEnv(s);
    }

    std::printf("%s: %d checks, %d failed\n", fails == 0 ? "PASS" : "FAIL", checks, fails);
    return fails == 0 ? 0 : 1;
}

File: python/tests/test_top_addressing.py
# -*- coding: utf-8 -*-
"""动态 name server（TopAddressing / DefaultTopAddressing）单测 —— 本地 mock HTTP server。

Java 语义锚点（common/namesrv/DefaultTopAddressing.java + MixAll.getWSAddr +
client/impl/MQClientAPIImpl.fetchNameServerAddr，5.5.1 逐条核对）：
  * WS 地址：``http://<domain>:8080/rocketmq/<subgroup>``；domain 自带端口（含 ':'）
    时不追加 :8080；
  * unitName 非空白 → URL 追加 ``-<unitName>?nofix=1``；para 非空 → ``?k=v&...``；
  * HTTP GET（超时 3000ms）code==200 → 响应体 ``clearNewLine``（trim 后截断到第一个
    \\r 或 \\n）作为 NS 地址串；否则 None；
  * ``fetchNameServerAddr``：地址**变化才应用**（按 ';' 切分更新地址表）；
  * ``MQClientInstance``：只在未配置静态地址时，start() fetch 一次 + 10s/2min 周期刷新。

mock HTTP server 用标准库 http.server 起在 127.0.0.1 随机端口，不依赖外网。
"""
from __future__ import annotations

import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

import pytest

from rocketmq.client.consumer import DefaultMQPushConsumer
from rocketmq.client.mq_client import MQClientInstance
from rocketmq.client.top_addressing import DefaultTopAddressing, clear_new_line


# ------------------------------------------------------------------ mock server

class _MockAddrServer:
    """极简地址服务器：可编程返回状态码与响应体。"""

    def __init__(self):
        self.status = 200
        self.body = "127.0.0.1:9876"
        self.hit_paths: list = []
        self._server = ThreadingHTTPServer(("127.0.0.1", 0), self._make_handler())
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()

    @property
    def port(self) -> int:
        return self._server.server_address[1]

    def _make_handler(self):
        outer = self

        class H(BaseHTTPRequestHandler):
            def do_GET(self):  # noqa: N802
                outer.hit_paths.append(self.path)
                body = outer.body.encode("utf-8")
                self.send_response(outer.status)
                self.send_header("Content-Length", str(len(body)))
                self.end_headers()
                self.wfile.write(body)

            def log_message(self, *args):  # 静音
                pass

        return H

    def stop(self):
        self._server.shutdown()
        self._server.server_close()


@pytest.fixture()
def addr_server():
    s = _MockAddrServer()
    yield s
    s.stop()


def _top_for(server: _MockAddrServer, **kw) -> DefaultTopAddressing:
    """指向 mock server 的 TopAddressing（domain 带端口 → 不追加 :8080）。"""
    return DefaultTopAddressing(domain="127.0.0.1:%d" % server.port, **kw)


# ------------------------------------------------------------------ URL 构造

class TestWsAddr:
    def test_default_port_appended(self):
        # Java MixAll.getWSAddr：domain 无端口 → 补 :8080
        assert DefaultTopAddressing.get_ws_addr("jmenv.tbsite.net") == \
            "http://jmenv.tbsite.net:8080/rocketmq/nsaddr"

    def test_domain_with_port_skips_default(self):
        assert DefaultTopAddressing.get_ws_addr("host:12345") == \
            "http://host:12345/rocketmq/nsaddr"

    def test_custom_subgroup(self):
        assert DefaultTopAddressing.get_ws_addr("h", "mygrp") == \
            "http://h:8080/rocketmq/mygrp"


class TestBuildUrl:
    def test_plain(self):
        ta = DefaultTopAddressing(ws_addr="http://h:8080/rocketmq/nsaddr")
        assert ta.build_url() == "http://h:8080/rocketmq/nsaddr"

    def test_unit_name(self):
        ta = DefaultTopAddressing(ws_addr="http://h:8080/rocketmq/nsaddr", unit_name="unitA")
        assert ta.build_url() == "http://h:8080/rocketmq/nsaddr-unitA?nofix=1"

    def test_para_only(self):
        ta = DefaultTopAddressing(ws_addr="http://h:8080/rocketmq/nsaddr",
                                  para={"k1": "v1", "k2": "v2"})
        url = ta.build_url()
        assert url.startswith("http://h:8080/rocketmq/nsaddr?")
        assert "k1=v1&k2=v2" in url or "k2=v2&k1=v1" in url

    def test_unit_name_with_para(self):
        # Java：unitName + para → "-unit?nofix=1&k=v..."（末尾 & 去掉）
        ta = DefaultTopAddressing(ws_addr="http://h:8080/rocketmq/nsaddr",
                                  unit_name="u", para={"k": "v"})
        assert ta.build_url() == "http://h:8080/rocketmq/nsaddr-u?nofix=1&k=v"

    def test_blank_unit_name_is_ignored(self):
        ta = DefaultTopAddressing(ws_addr="http://h:8080/rocketmq/nsaddr", unit_name="   ")
        assert ta.build_url() == "http://h:8080/rocketmq/nsaddr"


# ------------------------------------------------------------------ clearNewLine

class TestClearNewLine:
    def test_trim_and_cut_at_cr(self):
        assert clear_new_line("  1.2.3.4:9876\r\nrest") == "1.2.3.4:9876"

    def test_cut_at_lf(self):
        assert clear_new_line("a:9876\nb:9877") == "a:9876"

    def test_no_newline(self):
        assert clear_new_line("  a:9876  ") == "a:9876"

    def test_empty(self):
        # Java trim() 会把前导 \r\n 一并去掉，所以这里 trim 后就是 "x"（与 Java 一致）
        assert clear_new_line("   \r\n x") == "x"
        assert clear_new_line("\r\n") == ""


# ------------------------------------------------------------------ 取址行为

class TestFetchNsAddr:
    def test_200_returns_body(self, addr_server):
        addr_server.body = "10.0.0.1:9876;10.0.0.2:9876\nextra"
        ta = _top_for(addr_server)
        assert ta.fetch_ns_addr(verbose=False) == "10.0.0.1:9876;10.0.0.2:9876"
        # URL 走的是 /rocketmq/<subgroup>
        assert addr_server.hit_paths[0].startswith("/rocketmq/nsaddr")

    def test_non_200_returns_none(self, addr_server):
        addr_server.status = 500
        ta = _top_for(addr_server)
        assert ta.fetch_ns_addr(verbose=False) is None

    def test_connection_error_returns_none(self):
        # 没有服务监听的端口 → 连接失败 → None（Java catch IOException）
        ta = DefaultTopAddressing(domain="127.0.0.1:1", timeout_millis=300)
        assert ta.fetch_ns_addr(verbose=False) is None

    def test_no_domain_means_disabled(self):
        ta = DefaultTopAddressing()          # 未给 domain、env 未设
        assert ta.ws_addr == ""
        assert ta.fetch_ns_addr() is None

    def test_unreachable_at_start_is_reported(self):
        ta = DefaultTopAddressing(domain="127.0.0.1:1", timeout_millis=300)
        # verbose=True 走 error 日志但不抛（Java 语义）
        assert ta.fetch_ns_addr(verbose=True) is None


class TestFetchAndApply:
    def test_applies_change_only(self, addr_server):
        ta = _top_for(addr_server)
        assert ta.fetch_and_apply() == "127.0.0.1:9876"      # 第一次：变化 → 应用
        assert ta.fetch_and_apply() is None                  # 第二次：相同 → 不应用
        addr_server.body = "10.0.0.9:9876"
        assert ta.fetch_and_apply() == "10.0.0.9:9876"       # 变了 → 应用

    def test_blank_body_not_applied(self, addr_server):
        addr_server.body = "   "
        ta = _top_for(addr_server)
        assert ta.fetch_and_apply() is None


# ------------------------------------------------------------------ 实例集成

class TestMQClientInstanceIntegration:
    def test_start_with_static_addrs_never_fetches(self, addr_server):
        ta = _top_for(addr_server)
        mqc = MQClientInstance("c@dyn", ["127.0.0.1:9876"])
        mqc.top_addressing = ta
        mqc.start()
        try:
            assert addr_server.hit_paths == []               # 配了静态地址就不该问地址服务器
        finally:
            mqc.shutdown()

    def test_start_with_empty_addrs_fetches_once(self, addr_server):
        ta = _top_for(addr_server)
        mqc = MQClientInstance("c@dyn2", [])
        mqc.top_addressing = ta
        mqc.start()
        try:
            assert mqc.name_server_addrs == ["127.0.0.1:9876"]
            assert len(addr_server.hit_paths) == 1
        finally:
            mqc.shutdown()

    def test_start_fails_when_address_server_returns_none(self, addr_server):
        addr_server.status = 500
        mqc = MQClientInstance("c@dyn3", [])
        mqc.top_addressing = _top_for(addr_server)
        with pytest.raises(Exception):
            mqc.start()

    def test_periodic_refresh_picks_up_change(self, addr_server):
        ta = _top_for(addr_server)
        mqc = MQClientInstance("c@dyn4", [])
        mqc.top_addressing = ta
        mqc.start()
        try:
            addr_server.body = "10.0.0.9:9876"
            mqc.fetch_name_server_addr()
            assert mqc.name_server_addrs == ["10.0.0.9:9876"]
        finally:
            mqc.shutdown()


# ------------------------------------------------------------------ 消费者守卫

class TestConsumerGuard:
    def test_empty_addrs_rejected_without_domain(self, monkeypatch):
        monkeypatch.delenv("ROCKETMQ_NAMESRV_DOMAIN", raising=False)
        c = DefaultMQPushConsumer("GID_DynNsUnit")
        c.subscription_data = {"T": None}  # 只为让守卫先卡在地址上
        with pytest.raises(Exception, match="name server address"):
            c.start()

    def test_empty_addrs_allowed_with_domain_env(self, addr_server, monkeypatch):
        monkeypatch.setenv("ROCKETMQ_NAMESRV_DOMAIN", "127.0.0.1:%d" % addr_server.port)
        c = DefaultMQPushConsumer("GID_DynNsUnit")
        c.name_server_addrs = []
        # 守卫放行后走到 MQClientInstance.start：地址服务器返回的地址可用 → 不抛
        mqc = MQClientInstance("c@guardenv", [])
        mqc.top_addressing = DefaultTopAddressing()
        mqc.start()
        try:
            assert mqc.name_server_addrs == ["127.0.0.1:9876"]
        finally:
            mqc.shutdown()
